tlair/src/main/scala/at/forsyte/apalache/tla/lir/TlaLevel.scala

```diff
@@ -0,0 +1,93 @@
+package at.forsyte.apalache.tla.lir
+
+/**
+ * The level of a TLA+ expression, as explained in: Lamport. Specifying Systems, p. 322.
+ */
+sealed trait TlaLevel extends Ordered[TlaLevel] {
+
+  /**
+   * The level number as assigned in the book.
+   */
+  val level: Int
+
+  /**
+   * A textual representation of the level.
+   */
+  val name: String
+
+  /**
+   * Compare this level to another level.
+   *
+   * @param that the level to compare
+   * @return 0, if the levels are equal; a negative number if `this` is smaller than `that`, and a positive number
+   *         if `this` is larger than `that`.
+   */
+  override def compare(that: TlaLevel): Int = {
+    level - that.level
+  }
+
+  /**
+   * Join the levels by computing the maximum of the levels of `this` and `that`.
+   *
+   * @param that the level to join with
+   * @return the minimal level j that satisfies: `j <= this` and `j <= that`.
+   */
+  def join(that: TlaLevel): TlaLevel = {
+    TlaLevel.fromInt(Math.max(level, that.level))
+  }
+
+  /**
+   * Join the level with a sequence of level.
+   *
+   * @param otherLevels a sequence of levels
+   * @return the join of `this` and otherLevels
+   */
+  def join(otherLevels: Seq[TlaLevel]): TlaLevel = {
+    otherLevels.foldLeft(this) { case (l, r) => l.join(r) }
+  }
+}
+
+object TlaLevel {
+  private val INT_TO_LEVEL = Seq(TlaLevelConst, TlaLevelState, TlaLevelAction, TlaLevelTemporal)
+
+  def fromInt(levelNo: Int): TlaLevel = {
+    INT_TO_LEVEL.find(_.level == levelNo) match {
```

Comment:
Minor gripe,I would use an indexed collection and do boundary tests here, instead of find.

Reply:
I've replaced it with find, which should be ok for a list of 4

Reply:
Ha-ha. My solution was exactly as you suggest, and @shonfeder did not like it :P